File: templates/java_rest_api_template/src/main/java/com/template/restapi/resources/cache/RedisCacheService.java
/**
 * #############################################################################
 * ### Redis cache service
 * ###
 * ### @file RedisCacheService.java
 * ### @author Sebastian Russo
 * ### @date 2026
 * #############################################################################
 *
 * Application-level cache helpers. Controllers and services should use this
 * class instead of importing {@link RedisClient} directly.
 */
package com.template.restapi.resources.cache;

import java.util.Optional;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

import org.springframework.stereotype.Service;

@Service
public class RedisCacheService {

    private final RedisClient redisClient;
    private final ObjectMapper objectMapper;

    public RedisCacheService(RedisClient redisClient, ObjectMapper objectMapper) {
        this.redisClient = redisClient;
        this.objectMapper = objectMapper;
    }

    public <T> Optional<T> cacheGet(String key, Class<T> type) {
        String raw = redisClient.get(key);
        if (raw == null) {
            return Optional.empty();
        }
        try {
            return Optional.of(objectMapper.readValue(raw, type));
        } catch (JsonProcessingException ex) {
            return Optional.empty();
        }
    }

    public boolean cacheSet(String key, Object value, int expirationSeconds) {
        try {
            String json = objectMapper.writeValueAsString(value);
            return redisClient.setEx(key, expirationSeconds, json);
        } catch (JsonProcessingException ex) {
            return false;
        }
    }

    public boolean cacheDelete(String key) {
        return redisClient.delete(key);
    }
}

File: templates/csharp_rest_api_template/src/RestApiTemplate/Resources/Cache/RedisCacheService.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Template.RestApi.Resources.Cache;

/// <summary>
/// Application-level cache helpers. Controllers and services should use this class
/// instead of importing <see cref="RedisClient"/> directly.
/// </summary>
public sealed class RedisCacheService(RedisClient redisClient)
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
    };

    public T? CacheGet<T>(string key)
    {
        var raw = redisClient.Get(key);
        if (raw is null)
        {
            return default;
        }

        try
        {
            return JsonSerializer.Deserialize<T>(raw, JsonOptions);
        }
        catch (JsonException)
        {
            return default;
        }
    }

    public bool CacheSet<T>(string key, T value, int expirationSeconds)
    {
        try
        {
            var json = JsonSerializer.Serialize(value, JsonOptions);
            return redisClient.SetEx(key, expirationSeconds, json);
        }
        catch (NotSupportedException)
        {
            return false;
        }
    }

    public bool CacheDelete(string key) => redisClient.Delete(key);
}
